File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.ans = []
        self.dfs(nums, [False] * len(nums), [])
        return self.ans
    
    def dfs(self, nums, taken, ls):
        all_taken = True
        for i in range(len(nums)):
            if not taken[i]:
                all_taken = False
                taken[i] = True
                ls.append(nums[i])
                self.dfs(nums, taken, ls)
                ls.pop()
                taken[i] = False
        if all_taken:
            self.ans.append(ls.copy())

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        function<void(vector<int>&, int)> get = [&](vector<int>& cur, int taken) {
            if (cur.size() == nums.size()) {
                ans.push_back(cur);
                return;
            }
            for (int i=0; i<nums.size(); i++) {
                if (!((1 << i) & taken)) {
                    cur.push_back(nums[i]);
                    get(cur, taken | (1 << i));
                    cur.pop_back();
                }
            }
        };
        vector<int> tmp;
        get(tmp, 0);
        return ans;
    }
};